### PR TITLE
Convert external source distributions to C++

### PR DESCRIPTION
**Pull request by paulromano**
In this branch, I've converted the `SourceDistribution` type to C++. In addition to removing source.F90 and source_header.F90, I was also able to get rid of distribution_univariate.F90 and distribution_multivariate.F90 since source distributions were the last thing that depended on them that hadn't been converted. Net reduction of about 1k lines of Fortran.

### CODE REVIEW

File: src/source.cpp [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+
+  // Determine total source strength
+  double total_strength = 0.0;
+  for (auto& s : external_sources)
+    total_strength += s.strength();
+
+  // Sample from among multiple source distributions
+  auto n_source = external_sources.size();
+  int i = 0;
+  if (n_source > 1) {
```
Comment (smharper) [inline src/source.cpp]: `n_source` is only used here so may as well test `external_sources.size() > 1`

File: src/source.cpp [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    double xi = prn()*total_strength;
+    double c = 0.0;
+    for (i = 0; i < external_sources.size(); ++i) {
```
Comment (smharper) [inline src/source.cpp]: I found it slightly confusing that `i` was initialized both here and above.  Maybe just
```C++
for (; i < external_sources.size(); ++i) {
```
for clarity.  (Or just declare `int i;` above)

File: src/source.cpp [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+  if (path_source.empty()) {
+    // Get pointer to source bank
+    Bank* source_bank;
+    int64_t n;
+    openmc_source_bank(&source_bank, &n);
+
+    for (int64_t i = 0; i < openmc_work; ++i) {
+      // initialize random number seed
+      int64_t id = (openmc_total_gen + overall_generation())*n_particles +
+        work_index[openmc::mpi::rank] + i + 1;
```
Comment (smharper) [inline src/source.cpp]: I've been operating under the assumption that we should indent continued lines 5 spaces like we did in Fortran, although I see the style guide doesn't mention it.  Do you have an opinion on 2 vs. 5 spaces for this indentation?

Comment (paulromano) [reply src/source.cpp]: The five space thing was a default continuation line indent in emacs for Fortran, so I would actually not propose adopting that convention going forward. Two spaces seems sufficient.

Comment (smharper) [reply src/source.cpp]: Agreed, 2 spaces seems fine.

**Pull request merged**